### ISSUE TITLE
Question regarding weathercodes

### ISSUE THREAD

**Issue by woheller69** [open-meteo/open-meteo#228]
Is there a special reason behind showing the worst case weathercode in daily values?
On a day with some fog in the morning and otherwise sun all day, fog will be shown as code for the day which does not give a realistic picture of the day.

**Comment by woheller69** [comment]
I think about reevaluating the hourly codes in my app between sunrise and sunset.
And if there is sun in the weather codes of at least 2 hours I could map

overcast clouds -> scattered clouds
rain -> shower rain

etc.

**Comment by woheller69** [comment]
_hours of sun_ might also be helpful

**Comment by Dany-Dan** [comment]
> Would maybe adding some sort of text summary on the daily blocks help at all or would it be too complex? I know most other APIs have some sort of text summary describing the day in more detail and would be a nice addition.

Personally, I prefer a language file based on weathercodes to be integrated by the developer in his application. It is more flexible and editable by the developer. And also less restrictive for Patrick.
Of course, it is necessary to create it!
So I created mine which contains 6 different languages. But except French (I'm French) and English, the rest is a translation via Google and DeepL...
If Patrick agrees, I could post it here or in the Discussion section and together we could improve the translation.
And later, Patrick could post it on the Documentation page of his site.

**Comment by patrick-zippenfenig** [comment]
Sure, table for the weather code documentation could be extended by multiple languages.

If weather codes are used to describe daily weather, there will be many more cases. E.g. codes could be defined for "mostly sunny, but rain in the evening". For a good user experience it might be better to always analyse the hourly weather code and try create small text snippets and explain transitions.

**Comment by woheller69** [comment]
On a day with 10 hours of sun and one hour of rain I think it is more important to have sun protection than an umbrella...
I would use a weather symbol that also includes the worst case, such as a shower symbol with rain, clouds, and sun, instead of just a rain symbol for that day.

**Comment by woheller69** [comment]
I included some code in my app to reanalyze hourly codes between sunset and sunrise and change codes if needed.
Maybe someone wants to try.

https://github.com/woheller69/omweather/issues/3

**Comment by patrick-zippenfenig** [comment]
The timeframe is 0-23 hour. If there is snow forecast at 7 or 8 pm, the weather code should include it, unless it is very light snow of <0.1 mm water (<0.7 cm snow depth). 

Weather codes are ordered by severity. For the daily weather code, the highest value from hourly values is returned. This is also the reason, why morning fog, can dominate the daily weather code, although it is perfectly sunny throughout the day.

**Comment by DadiBit** [comment]
> Sure, table for the weather code documentation could be extended by multiple languages.
> 
> If weather codes are used to describe daily weather, there will be many more cases. E.g. codes could be defined for "mostly sunny, but rain in the evening". For a good user experience it might be better to always analyse the hourly weather code and try create small text snippets and explain transitions.

Regarding this point, I think a way to solve it "easily" could be by using a param in the query like `&weathercode_worst=false`* and return an `Int` that contains multiple statuses operations that are shifted with bitwise operations (keeping under the 16 or 32 bit shifts). Considering that right now the wmo codes are from 1 to 99, we might give them a "space" of 7 bits (up to 128-1) or 8 bits (up to 256-1). Let's actually suppose we pick the second option: a single wmo code would occupy a full byte, which is much easier to extract than a 7-bits `Int`. Let's say that we use 32 bits integers: we have 4 byte, thus 4 codes.
Now, let me link to issue #84: let's say that:
- `morning` 4:00-12:00
- `afternoon` 12:00-16:00,
- `evening` 16:00-20:00
- `night` 20:00-4:00 (of the next day? Idk, I'm just proposing a possible solution :smile: )
We could have that the first (Endianess matters! But LE is better, since it's the default on Linux and I believe that the api can be mostly used by phones and Linux machines) byte is `morning`, the second `afternoon` and so on....

The biggest drawbacks of using bit shifts are 1) that values can't be of more than 4 bytes: it's better not to use more than 32 bits (eg: JavaScript can work with 32 bits maximum on bitwise operations) 2) that requires a bit of engegnieering to get it working (it's not easy to deal with endianess with some programming languages or on some machines)
The main advantage would be that the codes would still be `Int`s, easing api library integration (it should be just as easy as implementing a few lines of code, in theory: `code & 0xff000000` & Co.)

**Comment by Leftium** [comment]
> Weather codes are ordered by severity.

@patrick-zippenfenig Do you have a source for this? While higher WMO code generally indicates higher severity, I don't think this is always true. For example: WMO 65 (heavy intensity rain) seems more severe than WMO 80 (slight rain showers).

I have come up with a "periodic table" of WMO codes, where the more severe conditions are placed farther to the right and/or lower:

![image](https://github.com/open-meteo/open-meteo/assets/381217/e85a38b9-4998-4215-b55d-c57a9a1317a7)

Basically "showers," icy drizzle," "snow showers," (and probably "snow grains") were placed out of order in terms of severity.

Also note the color-coding in the chart above groups the codes into categories, in order of increasing severity:
1. No precipitation
2. Rain
3. Mixed rain + snow
4. Snow
5. Thunderstorm

Categories 1-4 can be each organized into their own 3x3 "severity grid."
